### PR TITLE
Improve kube-proxy updating to always get the latest version&build #4344

### LINKED ISSUE #4344: Improve kube-proxy updating to always get the latest version&build
**What feature/behavior/change do you want?**
I want to be able to update my kube-proxy to the latest versions using `eksctl utils update-kube-proxy`. Currently we update it to `major.minor.patch.build.1` where the major/minor/patch mirror your kubernetes version. Example:
- k8s cluster 1.21.1
- `eksctl utils update-kube-proxy` would update `kube-proxy` to have image version `1.21.1.build.1`

## Problem
We hardcode in `build.1` into our update logic. See https://github.com/weaveworks/eksctl/pull/4091 for more context. The TLDR is that we don't have a way of discovering iof a `build.X` exists, but we know for sure that `build.1` exists, so we default to that. Its our "best effort" attempt to update the addon to the latest version.

## Fix
We can update our logic to query the EKS Addons API to discover what the latest image version is. For example we already have sub command for querying:
```
eksctl utils describe-addon-versions --name kube-proxy --kubernetes-version 1.19
2021-10-15 10:45:58 [ℹ]  describing addon versions for addon: kube-proxy
{
  Addons: [{
      AddonName: "kube-proxy",
      AddonVersions: [{
          AddonVersion: "v1.19.6-eksbuild.2",
          Architecture: ["amd64","arm64"],
          Compatibilities: [{
              ClusterVersion: "1.19",
              DefaultVersion: true,
              PlatformVersions: ["eks.4+"]
            }]
        },{
          AddonVersion: "v1.18.8-eksbuild.1",
          Architecture: ["amd64","arm64"],
          Compatibilities: [{
              ClusterVersion: "1.19",
              DefaultVersion: false,
              PlatformVersions: ["eks.4+"]
            }]
        }],
      Type: "networking"
    }]
}
```

In the above we can see that `v1.19.6-eksbuild.2` is the latest kube-proxy version for `1.19`, but if you were to run `eksctl utils update-kube-proxy` against a `1.19` cluster we would only update it to `v1.19.6-eksbuild.1`. 

Lets update our code to query the addons API to discover the latest version, ensuring that we are always using the latest version.

##  Caveats
The addons API only exists for `1.18` and newer clusters, so we will have to maintain support for the old code path for older clusters.

## Credit
thanks @suket22 for the suggestion :smile:

### PR DESCRIPTION
**Pull request by aclevername**
### Description

Closes #4344

`eksctl utils update-kube-proxy` now queries the EKS addons api (for clusters >=1.18) to check to see if it reports a newer kube-proxy version than the default (which is the `major.minor.patch-eksbuild.1` using the k8s version). Sometimes it reports a more up-to-date version, sometimes not. It uses which ever is more up-to-date. 

## Reviewing suggestion
The 2nd commit `query EKS API to discover latest kube-proxy version` is the one that implements the fix. The rest of the commits are refactoring the existing code. I recommend looking at the commit in isolation and then the rest of the PR.

## Manual test
Before this change on an 1.21 cluster that was upgraded from 1.20:
```
k -n kube-system get daemonsets.apps kube-proxy -o yaml | grep image
        image: <redacted>.dkr.ecr.eu-west-1.amazonaws.com/eks/kube-proxy:v1.20.7-eksbuild.1
        
eksctl utils update-kube-proxy --cluster jk-20 --region eu-west-1 --approve
2021-11-11 11:43:22 [ℹ]  eksctl version 0.67.0
2021-11-11 11:43:22 [ℹ]  using region eu-west-1
2021-11-11 11:43:22 [ℹ]  "kube-proxy" is now up-to-date

k -n kube-system get daemonsets.apps kube-proxy -o yaml | grep image
        image: <redacted>.dkr.ecr.eu-west-1.amazonaws.com/eks/kube-proxy:v1.21.2-eksbuild.1
```

After this change on a 1.21 cluster:
```
k -n kube-system get daemonsets.apps kube-proxy -o yaml | grep image
        image: <redacted>.dkr.ecr.eu-west-1.amazonaws.com/eks/kube-proxy:v1.20.7-eksbuild.1

./eksctl utils update-kube-proxy --cluster jk-20 --region eu-west-1 --approve
2021-11-11 11:55:49 [ℹ]  eksctl version 0.74.0-dev+d90c049e.2021-11-11T11:54:52Z
2021-11-11 11:55:49 [ℹ]  using region eu-west-1
2021-11-11 11:55:50 [ℹ]  "kube-proxy" is now up-to-date

k -n kube-system get daemonsets.apps kube-proxy -o yaml | grep image
        image: <redacted>.dkr.ecr.eu-west-1.amazonaws.com/eks/kube-proxy:v1.21.2-eksbuild.2
```

This improvement now means that users will be running `kube-proxy:v1.21.2-eksbuild.2` instead of `kube-proxy:v1.21.2-eksbuild.1` :smile: 

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/addons/default/addons.go [hunk 1]
```diff
@@ -1,27 +1,41 @@
...
 )
 
-func DoAddonsSupportMultiArch(clientSet kubernetes.Interface, rawClient kubernetes.RawClientInterface, controlPlaneVersion string, region string) (bool, error) {
-	kubeProxyUpToDate, err := IsKubeProxyUpToDate(clientSet, controlPlaneVersion)
+type AddonInput struct {
+	RawClient           kubernetes.RawClientInterface
+	EKSAPI              eksiface.EKSAPI
+	ControlPlaneVersion string
+	Region              string
+}
```
Comment (aclevername) [inline pkg/addons/default/addons.go:13]: kube proxy needs all 4
coredns needs 3
aws-node needs 2

either keep it as this or:
````

type Struct KubeProxyAddon {
	RawClient           kubernetes.RawClientInterface
	EKSAPI              eksiface.EKSAPI
	ControlPlaneVersion string
	Region              string}

type Struct KubeProxyAddon {
  	RawClient           kubernetes.RawClientInterface
	ControlPlaneVersion string
	Region              string
}


type Struct AWSNodeAddon {
  	RawClient           kubernetes.RawClientInterface
	Region              string
}


````

I don't mind :shrug:

File: pkg/addons/default/aws_node_test.go [hunk 2]
```diff
@@ -91,97 +98,98 @@ var _ = Describe("default addons - aws-node", func() {
...
+
+					awsNode, err := rawClient.ClientSet().AppsV1().DaemonSets(metav1.NamespaceSystem).Get(context.TODO(), da.AWSNode, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					//should be unchanged
+					Expect(awsNode.Spec).To(Equal(preUpdateAwsNode.Spec))
+				})
+			})
 		})
 	})
 })
```
Comment (aclevername) [inline pkg/addons/default/aws_node_test.go:195]: don't even attempt to review the diff, the previous test was a hot piece of garbage. Just checkout the file as a whole and see if the tests make sense plz :smile:

File: pkg/addons/default/kube_proxy_test.go [hunk 3]
```diff
@@ -2,44 +2,259 @@ package defaultaddons_test
...
+		})
+
+		When("it doesn't exist", func() {
+			BeforeEach(func() {
+				rawClient := testutils.NewFakeRawClient()
+				input.RawClient = rawClient
+				clientSet = rawClient.ClientSet()
+			})
+
+			// if it doesn't exist it doesn't need updating, so its up to date ¯\_(ツ)_/¯ according to #2667
```
Comment (aclevername) [inline pkg/addons/default/kube_proxy_test.go:73]: ¯\_(ツ)_/¯ according to #2667

File: pkg/addons/default/addons.go (hunk 1 above)
Comment (aclevername) [reply pkg/addons/default/addons.go:15]: or maybe just a struct for kube-proxy/coredns. Idk so many different ideas, I give up thinking.

File: pkg/addons/default/aws_node_test.go [hunk 4]
```diff
@@ -91,97 +98,97 @@ var _ = Describe("default addons - aws-node", func() {
...
+					awsNode, err := rawClient.ClientSet().AppsV1().DaemonSets(metav1.NamespaceSystem).Get(context.TODO(), da.AWSNode, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					//should be unchanged
+					Expect(awsNode.Spec).To(Equal(preUpdateAwsNode.Spec))
+				})
+			})
+
+			When("it doesn't need an update", func() {
+				BeforeEach(func() {
+					rawClient = testutils.NewFakeRawClient()
```
Comment (Skarlso) [inline pkg/addons/default/aws_node_test.go:163]: This before is repeated over and over again in other tests as well. Might be worth putting it into a function and calling it with different params?

File: pkg/addons/default/aws_node_test.go [hunk 5]
```diff
@@ -91,97 +98,97 @@ var _ = Describe("default addons - aws-node", func() {
...
+
+					awsNode, err := rawClient.ClientSet().AppsV1().DaemonSets(metav1.NamespaceSystem).Get(context.TODO(), da.AWSNode, metav1.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					//should be unchanged
+					Expect(awsNode.Spec).To(Equal(preUpdateAwsNode.Spec))
+				})
+			})
+
+			When("it doesn't need an update", func() {
+				BeforeEach(func() {
```
Comment (Skarlso) [inline pkg/addons/default/aws_node_test.go:162]: This before is repeated over and over again in other tests as well. Might be worth putting it into a function and calling it with different params?

**Review by Skarlso — APPROVED** [review]
Well, I look at this now for a loooong time, and I think it's fine. :D

Or at least I can't think of anything better hahahaha.

File: pkg/addons/default/kube_proxy_test.go [hunk 6]
```diff
@@ -2,44 +2,259 @@ package defaultaddons_test
...
+
+				It("returns an error", func() {
+					_, err := da.UpdateKubeProxy(input, false)
+					Expect(err).To(MatchError(ContainSubstring("no versions available for \"kube-proxy\"")))
+				})
+			})
+
+			When("there are no valid versions returned", func() {
+				BeforeEach(func() {
+					mockProvider.MockEKS().On("DescribeAddonVersions", &awseks.DescribeAddonVersionsInput{
```
[the remark this answers is not in the document]
Comment (aclevername) [reply pkg/addons/default/kube_proxy_test.go:216]: :thinking: I don't agree. The BeforeEach is useful for writing dry-code (not applicable in this scenario), but also for also separating out the test setup from the test itself. I quite like that the `It` contains all the assertions, the `BeforeEach` the stuff to setup the test. I feel like we use this pattern in most of eksctl (not to say its necessarily the correct thing though :sweat_smile: )

**Pull request merged**